### PR TITLE
feat: LSTM用の学習データ

### PR DESCRIPTION
**Pull request by maton369**

### CODE REVIEW

**Review by maton369 — COMMENTED** [review]
[GitHub Review] ch05-03: 日本語テキスト分類の前処理（5.3.1–5.3.4）

要約
- 本PRは Wikipedia 要約を用いた「歴史 vs 地理」の二値分類に向け、分かち書き（Janome）→語彙化（Tokenizer）→ID列化→パディング→学習/検証ラベル生成までの下ごしらえを行っている。
- 実装は概ね最短経路で組まれているが、学習理論・データリーク回避・未知語耐性・再現性という観点で要改善点が複数ある。
- 核となる指摘は「語彙辞書は訓練集合のみに基づいて構築すべき」「未知語トークンの導入」「パディング/切り詰めの方針の明示」「事前学習埋め込みの活用」である。

主要論点（理論中心）

1) 語彙構築とデータリーク
- 現状 `tokenizer.fit_on_texts(list_all_x)` により、テスト（list3_x）を含む全データで語彙辞書を作成している。
- 語彙（単語→ID）をテストで観測した語にまで最適化すると「情報漏洩（data leakage）」となり、汎化性能の推定が楽観に偏る。
- 理論的には、訓練分布から誘導される語彙写像  f_train: 𝒱_train→ℕ  を固定し、評価時は f_train を適用すべきである。
  従って、fit は訓練集合（歴史+地理）に限定し、テキスト変換は同一辞書で学習/検証へ適用するのが正道である。

2) 未知語（OOV）と表現力の安定性
- 日本語は語彙の裾が重く未知語が頻発する。`Tokenizer` を OOV 未設定で使うと未登録語が欠落し系列長が不安定化する。
- 記号論的には語彙外語 w∉𝒱_train に対して射影 π(w)=⟨UNK⟩ を与える（OOV トークン）ことで、系列構造の保存と統計的頑健性を確保できる。
- したがって `Tokenizer(oov_token="<UNK>")` を要求する。

3) パディング/トランケーション方針（時系列整形）
- LSTM は前後文脈の順序情報を学習するため、切り詰め位置（pre/post）とパディング位置（pre/post）が学習ダイナミクスに影響する。
- 典型的には「post-padding / post-truncating」を採用し、系列先頭の語順を温存するほうが意味保持に有利なことが多い（先頭に定義や固有名詞が来やすい Wikipedia では特に）。
- また MAX_LEN は経験則では「95パーセンタイル」付近に設定し、情報喪失と計算コストのトレードオフを設計すべきである。分布確認（長さヒストグラム/分位点）を推奨する。

4) 事前学習 Word2Vec の利用と表現の初期化
- 現状は Embedding 層を暗黙にランダム初期化する前提であるが、直前セクションで学習済み Word2Vec（gensim KeyedVectors 等）を得ている。
- 表現学習理論の観点では、語彙→連続ベクトル空間への写像 φ: 𝒱→ℝ^d を意味的に初期化すると、低リソース学習で収束性・汎化が改善しやすい。
- 実務では「(a) 学習済みベクトルで埋め込み行列を初期化」「(b) 凍結して上位のみ学習 or 微調整（fine-tune）」の二択を比較評価する。

5) データ生成過程と分布の取り扱い
- Wikipedia summary は更新・長さ・ドメイン混入のゆらぎがある。学習理論では「IID 仮定からのズレ」を最小化すべく、収集時点の固定・キャッシュ・前処理一貫性（NFKC 正規化、数字正規化）を要求する。
- 分かち書きの一貫性（Janome固有の活用原形化/品詞判断）と Tokenizer のトークン境界は一致しているが、派生系（Subword/BPE/SentencePiece）を導入すると OOV 問題をさらに緩和できる。

6) ラベル設計と評価指標
- 今回は二値ラベル（歴史=0, 地理=1）。クラス不均衡やリークがあると単純精度が過大評価される。
- 検証では Stratified 分割、適切な性能指標（F1、ROC-AUC、PR-AUC）、混同行列を最低限提示すべきである。

7) 再現性とバージョン整合性
- すでに遭遇している `keras.preprocessing` 不在は Keras 3 仕様での破壊的変更に起因する。tensorflow.keras への切替 or keras-preprocessing の明示依存を固定し、requirements にピン止めすべきである。
- 乱数種（Python/NumPy/TensorFlow）を固定し、結果の再現可能性を担保する。

具体的アクション（高優先度）

- [必須] 語彙構築を学習データのみに限定する：
  - `tokenizer.fit_on_texts(list1_x + list2_x)` とし、`list3_x` には `texts_to_sequences` のみ適用。
- [必須] 未知語トークンを導入：
  - `tokenizer = Tokenizer(oov_token="<UNK>")`。
- [推奨] パディング/切り詰め方針の明示：
  - `pad_sequences(..., maxlen=MAX_LEN, padding="post", truncating="post")`。
- [推奨] MAX_LEN を統計的に決定：
  - 文章長の分位点（例：P95）を測定し設定根拠をコメント化。
- [推奨] 事前学習 Word2Vec から埋め込み行列を構築し初期化：
  - `embedding_matrix[w_id] = kv[word]`、未知語はゼロ or 正規乱数。埋め込みは凍結/微調整を比較。
- [推奨] Keras 依存の一本化：
  - `from tensorflow.keras.preprocessing.text import Tokenizer` ほか、要件を `requirements.txt` に固定。
- [推奨] データリーク防止と評価の厳密化：
  - StratifiedKFold もしくは train/val/test 分離、評価指標の拡充。

参考コード断片（パッチ方針の提示、疑似差分）
- これはレビューの補助であり、最小限の変更点のみを列挙する。

(1) Tokenizer と語彙学習域の修正
    tokenizer = Tokenizer(oov_token="<UNK>")
    tokenizer.fit_on_texts(list1_x + list2_x)  # 訓練のみ

    # 検証/テストは同一辞書で数値化
    sequence_test = tokenizer.texts_to_sequences(list3_x)
    sequence_test = pad_sequences(sequence_test, maxlen=MAX_LEN, padding="post", truncating="post")

    sequence_train = tokenizer.texts_to_sequences(list1_x + list2_x)
    sequence_train = pad_sequences(sequence_train, maxlen=MAX_LEN, padding="post", truncating="post")

(2) 文章長分布の確認（根拠づけ）
    lengths = [len(s.split()) for s in (list1_x + list2_x)]
    # 分位点（例）：np.percentile(lengths, [50, 90, 95, 99]) をログ出力し MAX_LEN を妥当化

(3) 事前学習 Word2Vec の埋め込み行列（任意）
    # kv: gensim KeyedVectors, word_index: Tokenizer 語彙
    embedding_matrix = np.zeros((num_words + 1, EMBEDDING_DIM), dtype="float32")
    for word, idx in word_index.items():
        if word in kv:
            vec = kv[word]
            if vec.shape[0] == EMBEDDING_DIM:
                embedding_matrix[idx] = vec

    # Keras Embedding 初期化（trainable は実験で比較）
    Embedding(input_dim=num_words + 1, output_dim=EMBEDDING_DIM, weights=[embedding_matrix],
              input_length=MAX_LEN, trainable=False)

総評
- データ前処理の骨格は妥当であるが、評価の厳密性（リーク回避・未知語処理・長さ方針）と表現学習の活用（事前学習埋め込み）が未整備である。
- 上述の修正を適用すれば、モデルの汎化性能と再現性が理論・実務両面で改善する見込みである。
- したがって本PRは「Request changes」。最低限「語彙の学習域の限定」と「OOV 導入」を反映のうえ、再評価結果（指標/混同行列）を追記されたい。
